### ISSUE TITLE
torch.split / torch.chunk cause "too complex strides" for Inductor + CUDA graph

### ISSUE THREAD

**Issue by yf225** [pytorch/torchdynamo#1620]
When a model uses output of `torch.split` / `torch.chunk` ops as input, or has those ops internally, `torch._debug_has_internal_overlap` can complain about the size + strides of those tensors being too hard to determine whether memory overlap exists, preventing enablement of CUDA graph:

## Example 1: using output of `torch.split` as input to graph
```python
all_embs = torch.randn(8, 101168)
emb_split = [98400, 340, 40, 328, 380, 1320, 360]
split_emb = torch.split(all_embs, emb_split, dim=1)
```
`split_emb` now has tensors of size and stride:
```
torch.Size([8, 98400])
(101168, 1)
torch.Size([8, 340])
(101168, 1)
torch.Size([8, 40])
(101168, 1)
torch.Size([8, 328])
(101168, 1)
torch.Size([8, 380])
(101168, 1)
torch.Size([8, 1320])
(101168, 1)
torch.Size([8, 360])
(101168, 1)
```
Passing `split_emb` as input to "Inductor + CUDA graph" backend will cause CUDA graph to be disabled.

## Example 2: using `torch.chunk` within a graph
```python
_2384 = torch.randn(8, 4096)
chunk_list = torch.chunk(_2384, 16, dim=1)
```
`chunk_list` now has 16 tensors of same size and stride:
```
torch.Size([8, 256])
(4096, 1)
```
If we do this `torch.chunk` op within the graph and apply "Inductor + CUDA graph" backend, CUDA graph will be disabled.

cc. @ngimel

**Comment by yf225** [comment]
More cases where we want to know if can be considered "non-overlapped":
```python
dense = torch.zeros(64, 32)
assert not torch._debug_has_internal_overlap(dense)
assert not torch._debug_has_internal_overlap(dense.t())

strided = dense.split(4, dim=1)
assert not torch._debug_has_internal_overlap(strided[0])
assert not torch._debug_has_internal_overlap(strided[0].t())

unsqueezed = dense.unsqueeze(1)
assert not torch._debug_has_internal_overlap(unsqueezed)
assert not torch._debug_has_internal_overlap(unsqueezed.permute(1, 2, 0))

expanded = unsqueezed.expand(-1, 2, -1)
assert not torch._debug_has_internal_overlap(expanded)
assert not torch._debug_has_internal_overlap(expanded.permute(1, 2, 0))

gathered = dense.index_select(0, torch.IntTensor([1, 0, 1]))
assert not torch._debug_has_internal_overlap(gathered)
assert not torch._debug_has_internal_overlap(gathered.t())
```

**Comment by jansel** [comment]
We should stop using `torch._debug_has_internal_overlap()` and write our own check that is more selective.  

Perhaps:
```
x.numel() != <last element ptr> - <first element ptr> + 1
```
Then switch our cudagraphs wrapper to just copy the underlying storage.

**Comment by yf225** [comment]
> `x.numel() != <last element ptr> - <first element ptr> + 1`

Wonder should it be `>`? Like if total # elements is larger than the storage size, we know there is overlap?

Alternatively, should we consider doing it the following way?
```cpp
MemOverlap has_overlap(TensorImpl* t) {
  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(t->layout() == kStrided);

  if (t->is_non_overlapping_and_dense()) {
    return MemOverlap::No;
  }

  auto strides = t->strides();
  auto sizes = t->sizes();

  // If we're worried about sorting overhead, we can return too hard
  // whenever we see stride[i + 1] < stride[i]
  std::vector<int64_t> indices(strides.size());
  std::iota(indices.begin(), indices.end(), 0);

  std::sort(indices.begin(), indices.end(), [&strides](int64_t i1, int64_t i2) {
    return strides[i1] < strides[i2];
  });

  for (const auto i : c10::irange(strides.size())) {
    int64_t prev_stride = (i == 0) ? 1 : strides[indices[i - 1]];
    int64_t prev_size = (i == 0) ? 1 : sizes[indices[i - 1]];
    if (strides[indices[i]] < prev_stride * prev_size) {
      return MemOverlap::Yes;
    }
  }

  return MemOverlap::No;
}
```
If we translate the above to Python and put in Dynamo (instead of in MemoryOverlap.cpp), would it work? (thanks Jongsoo @jspark1105 for writing this)

**Comment by jansel** [comment]
On further thought, think overlap should actually be fine for cudagraphs.  When we do the copy just do:
```
storage_size = <last_ptr>-<first_ptr>+1
as_strided(dst, (storage_size), (1)).copy_(as_strided(src, (storage_size), (1))
```

The problem only comes when storage_size is larger, in which case it could be wasteful as we will need to copy many unused elements.

**Comment by ngimel** [comment]
If we combine inputs belonging to the same storage, the number of wasted elements hopefully won't be large.

**Comment by yf225** [comment]
SG - to combine these two ideas:

```python
# Step 1: group input tensors by `.storage().data_ptr()`
...

# Step 2: for each input tensor group, we do:
src_storage = tensors[0].storage()
storage_size = src_storage.size()
dst_storage = torch.empty(storage_size, dtype=src_storage.dtype, device=src_storage.device)
as_strided(dst_storage, (storage_size), (1)).copy_(as_strided(src_storage, (storage_size), (1))

# Step 3: create dst tensors based on `dst_storage` and input tensor strides
...
```

Does this look good in general?

re. negative strides, should we just ban it because it's not well-supported in eager mode either? https://github.com/pytorch/pytorch/issues/39224

**Comment by ngimel** [comment]
Yes, looks good, you can assume there are no negative strides.

**Comment by eellison** [comment]
See also https://github.com/pytorch/torchdynamo/issues/1315

**Comment by ezyang** [comment]
> If we combine inputs belonging to the same storage, the number of wasted elements hopefully won't be large.

We shouldn't assume this. If you have a graph where an input is `big_tensor[0]`, the algorithm that @yf225 proposed will end up allocating space for `big_tensor`, which is REALLY BAD.

If we have a bunch of inputs that alias the same storage, and they're non-overlapping, the primary benefit to doing a single storage copy is that you only have to launch one kernel to copy the inputs into the CUDA graph region. But we can (and should) JIT compile the copy kernel into the CUDA graph to reduce launch overhead anyway, in which case there isn't any problem copying in a bunch of non-overlapping regions, they'll all get coalesced anyway.

If you have a bunch of overlapping inputs, now the calculation is more complicated, because if you copy them individually you might inflate the actual memory you use (e.g., imagine rolling a tensor with strides). It's not so easy to figure out what to do here, so you probably should just compute the min/max physical index into storage and copy only that.

**Comment by eellison** [comment]
Another issue is that we separately allocating storage for all inputs. Filed https://github.com/pytorch/torchdynamo/issues/1692 


> But we can (and should) JIT compile the copy kernel into the CUDA graph to reduce launch overhead 

I actually tried this on friday and couldn't measure any speedup even on a model with a bunch of graph breaks and thus cudagraph copies, but still probably worth landing.

**Comment by ezyang** [comment]
> I actually tried this on friday and couldn't measure any speedup even on a model with a bunch of graph breaks and thus cudagraph copies, but still probably worth landing.

You want a model with a lot of live inputs into the partial graph, if there's just one or two inputs (parameters and activations don't count on the inductor codepath, since their addresses get hard coded in) this doesn't help at all. We noticed this was a problem with `aot_cudagraphs`, since that path doesn't currently do the hard coded parameter/activation optimization yet.

**Comment by ngimel** [comment]
>No, I don't see why this would be expensive. For the split case, while you are looping through the sizes, you just need to check that the stride is greater than the minimum required stride for contiguity. And we're already looping through the tensor if the "is contiguous and non overlapping" test fails, so I don't see any impact to ATen.

We'd need to also sort the strides, we are not guaranteed sorted strides in `has_internal_overlap`

>We shouldn't assume this. If you have a graph where an input is big_tensor[0], the algorithm that @yf225 proposed will end up allocating space for big_tensor, which is REALLY BAD.

This is what would happen today, except we stop it in the `has_internal_overlap` check, because we copy preserving strides. @yf225's algorithm improves on that because today if there were 2 inputs,  `big_tensor[0]` and `big_tensor[1]`, and they happened to pass `has_internal_overlap` check we would allocate memory equivalent to 2 `big_tensor`s.

**Comment by ezyang** [comment]
> We'd need to also sort the strides, we are not guaranteed sorted strides in has_internal_overlap

OK. For this particular case we could hard-code it to work for standard order strides, or as a modest generalization, anything that suggest memory format would have worked on (which is cheap.)

> This is what would happen today, except we stop it in the has_internal_overlap check, because we copy preserving strides.

Err, I don't understand. big_tensor[0] would be contiguous, so it wouldn't trigger overlap check. The issue here doesn't have to do with internal overlap; it is views that are a subset of a big tensor.

**Comment by ngimel** [comment]
`big_tensor[0]` doesn't have to be contiguous (it is, if `big_tensor` is contiguous, but `big_tensor` could be permuted), if it was then indeed there would be no problem. But suppose we fixed `has_internal_overlap` to be smarter, and our inputs are `big_tensor[:, 0]` and `big_tensor[:, 1]`. We copy preserving strides, so the copy of `big_tensor[:, 0]` would need (almost) as much as full `big_tensor`, same for `big_tensor[:, 1]` - we just 2x'd the memory, unless we preserve storage sharing relationships in our copies also.

**Comment by ezyang** [comment]
Aha! We're talking about different situations. Let me propose an algorithm that works for both of the cases we're discussing.

1. Collect all tensor arguments that share the same storage (we will assume that distinct storages never overlap memory; but if you don't want to assume this we can do something equivalent with data ptr instead).
2. If there is only one tensor:
  a. If the tensor does not have internal overlap, copy the tensor contents to a contiguous buffer (which may be the smaller than the original tensor, if it wasn't contiguous). Inductor internally can assume that the tensor is contiguous for subsequent optimization.
  b. If the tensor has internal overlap, find the smallest and largest physical location on the storage that the tensor may access. Copy that range of the storage to a contiguous buffer (which will represent the new storage) and reconstruct the tensor from this tensor (setting `storage_offset` to 0). Inductor MUST NOT assume the tensor is contiguous.
3. If there are multiple tensors, we have to generalize the algorithm.
  a. If all the tensors do not overlap (internally or with each other), then they can be treated as individual tensor inputs without any special storage handling. We avoid allocating memory in the cuda graph for a case where the tensors point into a single big buffer, without overlapping, with lots of dead space between them.
  b. If the tensors overlap (internally or with each other), find the smallest and largest physical location on the storage that may be accessed by ANY of the tensors, reconstruct the storages from this tensor (modifying storage offset as necessary; it may not be 0).

**Comment by yf225** [comment]
> Inductor internally can assume that the tensor is contiguous for subsequent optimization.

Can we do this without worrying about original strides of the tensor? It will simplify things a lot if we can :)

**Comment by ngimel** [comment]
2a - by the time we do copies here, inductor has already run and generated the kernels with strided inputs, we can't change striding from underneath. Could we have done compacting before, during AOT and initial codegen? Maybe, but @chillee was telling me that we need to preserve actual strides and not just permutation. I'm willing to go back to drawing board on this, but we can't just change the striding in the cuda graphs copies and expect everything to work out.

**Comment by yf225** [comment]
Journalist summary of current issues:
1. There is only one input `big_tensor[0]`, and we allocate new storage of `big_tensor` size right now - Not sure if we can avoid this (unless we don't worry about strides matching the original input and allocate `big_tensor[0]` as a contiguous tensor).
2. There are multiple non-overlapping inputs `big_tensor[0] / [1] / [2]`, we allocate 3x `big_tensor` size right now - We can avoid this by detecting and allocating just the shared storage size.
3. There are multiple overlapping inputs `expanded[:, 0] / [:, 1]`, we ban this right now - Wonder if we can enable it.

**Comment by yf225** [comment]
In absence of an elegant solution, should we go with https://github.com/pytorch/pytorch/pull/87058 to allow all types of views, gated by a flag `torchinductor.config.triton.cudagraphs_support_complex_view_tensors` to make it an explicit opt-in?

**Comment by ezyang** [comment]
If you are looking for a point fix to unblock, I prefer just making the internal overlap test slightly smarter in C++ (assuming this is enough to actually entirely fix your problem.) For any complicated solution, we should figure out the right thing and fix it once.

**Comment by yf225** [comment]
Sounds great, so I can make a smarter overlap test and actually put it in Dynamo to not slow down the internal overlap test.

I think we still want to figure out the optimization for shared storage, because soon we could be hitting some models that push on the memory boundary.

**Comment by eellison** [comment]
Seems like this is fixed
